### DISCUSSION TITLE
ABM EV MESA

### DISCUSSION THREAD

**Discussion by JaslinRikkaz123** [mesa/mesa#1037] [Q&A] [answered]
Dear mesa users, 
I have run the mesa ABM-EV Model but it shows type error : unhashable type : 'list'  in space. py line 474 in place agent. 

https://github.com/sajansen1404/ABM_EV_MODEL
Please help me to reslove it
Thank you

**Comment by rht** [comment]
Can you post the full traceback? That information is not sufficient.

**Reply by JaslinRikkaz123** [reply]
![Uploading 1621389792909..jpg…]()
Please solve this issue i stuck this moment

**Reply by rht** [reply]
Your picture does not load properly. Can you reupload?

**Comment by JaslinRikkaz123** [comment]
Ok find the attachment

**Reply by rht** [reply]
There is nothing attached.

**Comment by JaslinRikkaz123** [comment]
[image: cmd.jpeg]

THANK YOU

**Reply by rht** [reply]
Still no picture. Can you just copy paste the traceback text?

**Comment by JaslinRikkaz123** [comment]
C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master>mesa runserver
Interface starting at http://127.0.0.1:8521
Socket opened!
{"type":"reset"}
{"type":"get_step","step":1}
ERROR:tornado.application:Uncaught exception GET /ws (127.0.0.1)
HTTPServerRequest(protocol='http', host='127.0.0.1:8521', method='GET',
uri='/ws', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\tornado\websocket.py",
line 647, in _run_callback
    result = callback(*args, **kwargs)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\mesa\visualization\ModularVisualization.py",
line 207, in on_message
    self.application.model.step()
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\model.py", line
186, in step
    self.schedule.step()
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\schedule.py",
line 57, in step
    self.step_breed(agent_class)
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\schedule.py",
line 73, in step_breed
    agent.step()
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\agents.py", line
576, in step
    self.move()
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\agents.py", line
110, in move
    self.moveEV()                                                       #
move and use battery
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\agents.py", line
415, in moveEV
    self.model.grid.move_agent(self,self.new_position)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\mesa\space.py",
line 334, in move_agent
    self._place_agent(pos, agent)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\mesa\space.py",
line 474, in _place_agent
    self.empties.discard(pos)
TypeError: unhashable type: 'list'

**Comment by JaslinRikkaz123** [comment]
ok find the traceback here.

C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master>mesa runserver
Interface starting at http://127.0.0.1:8521
Socket opened!
{"type":"reset"}
{"type":"get_step","step":1}
ERROR:tornado.application:Uncaught exception GET /ws (127.0.0.1)
HTTPServerRequest(protocol='http', host='127.0.0.1:8521', method='GET',
uri='/ws', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\tornado\websocket.py",
line 647, in _run_callback
    result = callback(*args, **kwargs)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\mesa\visualization\ModularVisualization.py",
line 207, in on_message
    self.application.model.step()
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\model.py", line
186, in step
    self.schedule.step()
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\schedule.py",
line 57, in step
    self.step_breed(agent_class)
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\schedule.py",
line 73, in step_breed
    agent.step()
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\agents.py", line
576, in step
    self.move()
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\agents.py", line
110, in move
    self.moveEV()                                                       #
move and use battery
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\agents.py", line
415, in moveEV
    self.model.grid.move_agent(self,self.new_position)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\mesa\space.py",
line 334, in move_agent
    self._place_agent(pos, agent)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\mesa\space.py",
line 474, in _place_agent
    self.empties.discard(pos)
TypeError: unhashable type: 'list'

**Reply by rht** [reply]
Can you replace this line https://github.com/sajansen1404/ABM_EV_MODEL/blob/master/EV/agents.py#L415 with ` self.model.grid.move_agent(self,tuple(self.new_position))` and see what happens?

**Comment by JaslinRikkaz123** [comment]
C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master>mesa runserver
Interface starting at http://127.0.0.1:8521
Socket opened!
{"type":"reset"}
{"type":"get_step","step":1}
c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\core\fromnumeric.py:3372:
RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\core\_methods.py:170:
RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
ERROR:tornado.application:Uncaught exception GET /ws (127.0.0.1)
HTTPServerRequest(protocol='http', host='127.0.0.1:8521', method='GET',
uri='/ws', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\tornado\websocket.py",
line 647, in _run_callback
    result = callback(*args, **kwargs)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\mesa\visualization\ModularVisualization.py",
line 207, in on_message
    self.application.model.step()
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\model.py", line
187, in step
    self.datacollector.collect(self)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\mesa\datacollection.py",
line 183, in collect
    self.model_vars[var].append(reporter(model))
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\model.py", line
61, in high_usage
    return np.percentile(np.array(CP_usage), 75)
  File "<__array_function__ internals>", line 5, in percentile
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\lib\function_base.py",
line 3732, in percentile
    return _quantile_unchecked(
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\lib\function_base.py",
line 3851, in _quantile_unchecked
    r, k = _ureduce(a, func=_quantile_ureduce_func, q=q, axis=axis, out=out,
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\lib\function_base.py",
line 3429, in _ureduce
    r = func(a, **kwargs)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\lib\function_base.py",
line 3967, in _quantile_ureduce_func
    x1 = take(ap, indices_below, axis=axis) * weights_below
  File "<__array_function__ internals>", line 5, in take
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\core\fromnumeric.py",
line 191, in take
    return _wrapfunc(a, 'take', indices, axis=axis, out=out, mode=mode)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\core\fromnumeric.py",
line 58, in _wrapfunc
    return bound(*args, **kwds)
IndexError: cannot do a non-empty take from an empty axes.

**Reply by rht** [reply]
That's because sometimes `CP_usage` in https://github.com/sajansen1404/ABM_EV_MODEL/blob/master/EV/model.py#L61 is an empty list. So you have to do
```
if len(CP_usage) > 0:
    return np.percentile(np.array(CP_usage), 75)
else:
    return 0
```
or something.

**Comment by JaslinRikkaz123** [comment]
mesa runserver
Interface starting at http://127.0.0.1:8521
Socket opened!
{"type":"reset"}
{"type":"get_step","step":1}
c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\core\fromnumeric.py:3372:
RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\core\_methods.py:170:
RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)

**Reply by rht** [reply]
That's runtime warning, not an error. But of course should be fixed if possible. So, is the simulation running properly?

**Comment by JaslinRikkaz123** [comment]
No, it only runs for one step. i have attached the image
[image: image.png]

**Reply by rht** [reply]
No image. Can you use github.com web interface to upload the image? Not email. Also, can you add `np.seterr(all='raise')` in https://github.com/sajansen1404/ABM_EV_MODEL/blob/master/EV/model.py#L17 and paste the error tracebacks here?

**Comment by JaslinRikkaz123** [comment]
mesa runserver
Traceback (most recent call last):
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\runpy.py", line 197,
in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\runpy.py", line 87,
in _run_code
    exec(code, run_globals)
  File "C:\Users\Jaslin
Shaleemkhan\AppData\Local\Programs\Python\Python39\Scripts\mesa.exe\__main__.py",
line 7, in <module>
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\click\core.py",
line 829, in __call__
    return self.main(*args, **kwargs)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\click\core.py",
line 782, in main
    rv = self.invoke(ctx)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\click\core.py",
line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\click\core.py",
line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\click\core.py",
line 610, in invoke
    return callback(*args, **kwargs)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\mesa\main.py",
line 33, in runserver
    exec(code, {}, {})
  File "run.py", line 1, in <module>
    from EV.server import server
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\server.py", line
6, in <module>
    from EV.model import EV_Model
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\model.py", line
18
    np.seterr(all='raise')
IndentationError: unexpected indent

**Reply by rht** [reply]
That's indentation error. Just indent your code properly.

**Comment by JaslinRikkaz123** [comment]
{"type":"get_step","step":1}
c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\core\fromnumeric.py:3372:
RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
ERROR:tornado.application:Uncaught exception GET /ws (127.0.0.1)
HTTPServerRequest(protocol='http', host='127.0.0.1:8521', method='GET',
uri='/ws', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\tornado\websocket.py",
line 647, in _run_callback
    result = callback(*args, **kwargs)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\mesa\visualization\ModularVisualization.py",
line 207, in on_message
    self.application.model.step()
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\model.py", line
198, in step
    self.datacollector.collect(self)
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\mesa\datacollection.py",
line 183, in collect
    self.model_vars[var].append(reporter(model))
  File "C:\Users\Jaslin Shaleemkhan\ABM_EV_MODEL-master\EV\model.py", line
27, in mean_all_battery
    B = np.mean(agent_battery_levels)
  File "<__array_function__ internals>", line 5, in mean
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\core\fromnumeric.py",
line 3372, in mean
    return _methods._mean(a, axis=axis, dtype=dtype,
  File "c:\users\jaslin
shaleemkhan\appdata\local\programs\python\python39\lib\site-packages\numpy\core\_methods.py",
line 170, in _mean
    ret = ret.dtype.type(ret / rcount)
FloatingPointError: invalid value encountered in double_scalars

**Comment by JaslinRikkaz123** [comment]
now it's working thank you for your help.

**Comment by rht** [comment, marked as answer]
@JaslinRikkaz123 this is a much cleaner fix to the repo: https://github.com/rht/ABM_EV_MODEL. I recommend you to use this instead of all the hacks I have told you earlier.

**Reply by rht** [reply]
Just for bookkeeping purpose, can you mark this discussion as resolved once you have confirmed that my fork works?

**Reply by JaslinRikkaz123** [reply]
How to do that?

**Reply by rht** [reply]
https://docs.github.com/en/discussions/managing-discussions-for-your-community/moderating-discussions#marking-a-comment-as-an-answer

**Reply by JaslinRikkaz123** [reply]
Done
